rfcs/006-package-management.md

```diff
@@ -0,0 +1,425 @@
+something to be present at a given path. In order to mitigate this:
+
+- Packages imported from the index are not allowed to have path dependencies.
+- Packages imported from git can have path dependencies, but only if they point
+  within the same git repo. The lock-file treats these path dependencies as though
+  they were git dependencies.
+
+Therefore your dependency tree can have a few path-dependency subtrees, but only
+at the root.
+
+## Stale lock files
+
+What happens if there's a lock-file, but someone modifies the manifest? The
+lock-file might need to be regenerated: certainly it might need some new
+entries, but also there might be new version conflicts that require a different
+resolution. In this case, we treat the lock-file as a suggestion instead of a
+hard constraint: during resolving, when choosing the next package version to
+try, we try to pick the locked version first. But if we run into a resolution
+conflict, we allow a different version to be chosen (and notify the user
+that a package was changed).
+
+This behavior is similar to what cargo does. It has the advantage that if the
+new manifest is compatible with the old lock-file, nothing will be changed.
+
+### Alternative: interactive prompt
+
+Instead of merely notifying the user that the lock-file changed, we could
+require them to approve the changes.
+
+### Alternative: explicit regeneration
+
+Instead of automatically updating the lock-file when the manifest changes,
+we could update it only on explicit commands: `plate eval` would use an
+old lock-file if it exists, while `plate lock` would read the manifest
+and update the lock-file. This alternative is potentially more efficient
+(as `plate eval` wouldn't need to re-read the manifest, and the manifests
+of path-dependencies, on every invocation), but makes it easy to accidentally
+use out-of-date packages.
+
+## Version compatibility and resolution
+
+How do we handle a package that gets imported multiple times in the dependency
+tree?
+
+For path and git dependencies, there isn't much choice. Dependencies from the
+registry are the most interesting.  Some languages (e.g. python) insist that
+each package resolves to a single version across the whole dependency tree.
+Other languages allow multiple versions, keeping track of which package in the
+dependency tree needs to import which version of a package.
+
+I think we want to allow multiple versions of a package; the alternative can be
+fragile and annoying. But then we need to figure out how many different versions
+to allow. There's a trade-off: if we allow pulling in a different version
+every time a package gets imported, solving the dependency graph is easy.
+But it increases the chance of getting incompatibilities at runtime: we might
+accidentally get a value from `util@1.1` and try to pass it to an incompatible
+function defined in `util@1.2`. Pulling in too many different versions also
+increases the total number of packages in the dependency graph.
```

Comment:
Just for reference, the Bazel bzlmod solution is to use [Go's MVS](https://research.swtch.com/vgo-mvs) and fail if they encounter incompatible version requests for the same package. But, the root package can declare [single or multi version overrides](https://bazel.build/external/module#overrides).

Reply:
That's a very interesting read, thanks! This approach and the one of allowing duplicate semver-compatible versions seem to be pretty much mutually exclusive, so I guess we need to pick one or the other?

Reply:
I guess the user controlled override is compatible with either approach. But, otherwise, yes, I think so.